proposals/buildfile.md

```diff
@@ -0,0 +1,118 @@
+# Jib CLI Buildfile Specification
+
+Specification for a buildfile describing building a container image. This buildfile can be
+used by the jib-cli to generate a container. It is translated directly into a buildplan and
+passed to the builder.
+
+```
+# "FROM" with option to use os/architecture for manifest lists
+baseImage: "gcr.io/distroless/java:8"
+osHint: linux
+architectureHint: amd64
```

Comment:
Why `Hint`?  It could signify something optional that could be ignored?  Why not `os` and `arch`?

Reply:
I just copied this part from the buildplan. @chanseokoh why did we use *hint* there?

Reply:
The user must provide a value to this field when the given base image is a manifest list. However, when the base image is an image, a tool must inherit the value from the base image while ignoring this field (or the tool may decided to throw an error).

As a general spec describing a plan to build an image in a declarative manner, I wanted to avoid possible confusion that someone mistakenly assumes that by specifying `os: arm64`, they will create an ARM64 image. Before you can actually check the base image reference at runtime, you won't know the final outcome.

Reply:
> However, when the base image is an image, a tool must inherit the value from the base image while ignoring this field (or the tool may decided to throw an error).

Where is this requirement come from?
 
Given that the os/arch is normally inherited, most users should never need to use this unless they're either creating a scratch image or if they mean to override the settings from a different image.  There are (were?) some ARM images on Docker Hub that identified as being amd64.  I believe distroless/base has no actual machine-specific code.

Reply:
> if they mean to override the settings from a different image.

Someone doing this doesn't make sense to me. It means an image has a wrong metadata, which should be fixed. I don't know if some Docker Hub images had wrong image metadata, but that's an error from the image creator.

In any case, this is a spec, and for this part, I think it doesn't have to mandate a certain interpretation in an implementation. A tool may provide an option like `--force-os-arch` (which I think is really unnecessary in practice.) But for the reason of the potential confusion I explained, I think `osHint` (or something else if there's a better term than `Hint`) is better than `os`

Reply:
Yeah I agree that correcting someone else's error is really our problem. But yeah I remember now that using *hint* here implies that you're not setting anything.